### PR TITLE
Grype facade string sbom

### PR DESCRIPTION
**Pull request by dspalmer99**
**What this PR does / why we need it**: Adds the ability to pass sboms to the grype facade as a string or a file path, and consequently updates utils to allow the execution of piped commands. Protects against requests to update the grype db where the new file does not exist. Adds a good number of unit tests.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>)(, fixes #<issue_number, ...)` format, will close the issue when PR is merged*: fixes #: 978

**Special notes**:

### CODE REVIEW

File: anchore_engine/utils.py
Comment (alfredodeza) [inline anchore_engine/utils.py:230]: in what situation the `cmd_lists` wouldn't be passed in? why is it useful to allow this?

File: tests/unit/anchore_engine/clients/test_grype_wrapper.py
Comment (alfredodeza) [inline tests/unit/anchore_engine/clients/test_grype_wrapper.py:102]: great use of pytest fixtures here!

File: tests/unit/anchore_engine/clients/test_grype_wrapper.py
Comment (alfredodeza) [inline tests/unit/anchore_engine/clients/test_grype_wrapper.py:227]: this is exactly the kind of thing that globals will cause. Now this test has effectively tainted the global variable, if this test were to fail, the next test that relies on this value could potentially fail. Figuring out which test did what and when to the global is time-consuming to say the least and all of this can be avoided by removing the global altogether

File: tests/unit/anchore_engine/test_utils.py
Comment (alfredodeza) [inline tests/unit/anchore_engine/test_utils.py:115]: thank you for adding these!

File: tests/unit/anchore_engine/test_utils.py
Comment (alfredodeza) [inline tests/unit/anchore_engine/test_utils.py:134]: I think it is very odd to have the function return `None, None, None` - this type of interaction with that function complicates how to deal with the output. I think it would be best to always return the expected values or raise an exception

File: tests/unit/anchore_engine/test_utils.py
Comment (alfredodeza) [inline tests/unit/anchore_engine/test_utils.py:149]: if this is catching any exception then at least the message should be verified, otherwise it might be too broad - making this test not that useful (false positive)

**Review by alfredodeza — CHANGES_REQUESTED** [review]

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:62]: Will do

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:234]: I've considered refactoring this whole module into a class so these state variables, currently implemented as globals, can just become data on the class. Not in scope for this change, but still something I would like to do. @nightfurys What do you think? We'd probably have to make sure it was created as a singleton during startup so that all of the users of this facade get the same instance.

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:246]: Will do

File: anchore_engine/utils.py
Comment (dspalmer99) [reply anchore_engine/utils.py:230]: As you said below, raising an exception is more apropriate. The only scenario is user error, the whole goal of protecting this block was to prevent an NPE on output.communicate()

Comment (dspalmer99) [reply anchore_engine/utils.py:230]: So I will make that change as well

File: tests/unit/anchore_engine/clients/test_grype_wrapper.py
Comment (dspalmer99) [reply tests/unit/anchore_engine/clients/test_grype_wrapper.py:102]: Thanks

File: tests/unit/anchore_engine/test_utils.py
Comment (dspalmer99) [reply tests/unit/anchore_engine/test_utils.py:115]: Np

File: tests/unit/anchore_engine/test_utils.py
Comment (dspalmer99) [reply tests/unit/anchore_engine/test_utils.py:149]: I will verify the message, but `Exception` is what the function under test raises, and it is used in enough places that I don't want to pull the testing involved on changing that into this change.

**Comment by nightfurys** [comment]
Few comments with reference to global variable access
- reduce the global variable read access by refactoring - create functions such as get_grype_db_session that access the global variable and return it. Would make swapping out with an other implementation if we decide to do that later
- move the global value writes to init functions for the same reason - set grype_db_session in __init_grype_db_engine. This is also consistent with how db connection is instantiated [engine](https://github.com/anchore/anchore-engine/blob/master/anchore_engine/db/entities/common.py#L148-L196) 

Circle CI is failing on the linter. I use black connect plugin in pycharm with linting on save option, makes linting a no-op for me

**Review by nightfurys — COMMENTED** [review]
Few comments with reference to global variable access
- reduce the global variable read access by refactoring - create functions such as get_grype_db_session that access the global variable and return it. Would make swapping out with an other implementation if we decide to do that later
- move the global value writes to init functions for the same reason - set grype_db_session in __init_grype_db_engine. This is also consistent with how db connection is instantiated [engine](https://github.com/anchore/anchore-engine/blob/master/anchore_engine/db/entities/common.py#L148-L196) 

Circle CI is failing on the linter. I use black connect plugin in pycharm with linting on save option, makes linting a no-op for me

**Comment by dspalmer99** [comment]
@nightfurys @alfredodeza I've made the changes you both requested, anything else in this one before I merge to the feature branch?

File: tests/unit/anchore_engine/clients/test_grype_wrapper.py
Comment (Vijay-P) [reply tests/unit/anchore_engine/clients/test_grype_wrapper.py:227]: @dspalmer99 do you think the globals could be refactored to a singleton? Or class variables? I really dislike the use of globals here.

File: tests/unit/anchore_engine/test_utils.py
Comment (alfredodeza) [reply tests/unit/anchore_engine/test_utils.py]: Yeah, if the underlying code raises `Exception` this is going to be more involved. Bummer :(

File: anchore_engine/clients/grype_wrapper.py
Comment (alfredodeza) [inline anchore_engine/clients/grype_wrapper.py:242]: I'm perhaps not understanding this well, but I'm not sure why `grype_db_dir` needs to be a global. Using a function here is definitely an improvement, but I can't reason why `global` is still being required

File: anchore_engine/clients/grype_wrapper.py
Comment (alfredodeza) [inline anchore_engine/clients/grype_wrapper.py:270]: excellent!

File: anchore_engine/utils.py
Comment (alfredodeza) [inline anchore_engine/utils.py:234]: This `else` is no longer needed here

File: anchore_engine/utils.py
Comment (alfredodeza) [inline anchore_engine/utils.py:236]: This is a nice way of writing a readable for-loop. `for book in books` is the usual example I give, and this implements it well. Thanks for writing it this way

**Review by nightfurys — APPROVED** [review]

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:302]: I had originally had some code in here (https://github.com/anchore/anchore-engine/blob/83cc719ada1469cab438d433f9b6e5796995418b/anchore_engine/clients/grype_wrapper.py#L255) that was going down the road of  allowing grype to be run more broadly, but took it out, primarily because I don't want to add maintenance and testing burden to the team for code we aren't actively using, but also because our use case here is that engine will build and manage grype db. If we also allow it to be run with latest grype db it pulls down itself that has a good chance of introducing confusion.

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:263]: I think this is a personal preference; to me it is more readable this way.

File: anchore_engine/clients/grype_wrapper.py
Comment (dspalmer99) [reply anchore_engine/clients/grype_wrapper.py:242]: Possibly I am not understanding the `global` correctly - my intention here (and previously when these variables were accessed elsewhere) was that when accessing these we always want the module-scoped `grype_db_dir` and `grype_db_session` declared above. This module differs from the other wrappers in that it needs to maintain state to be applied on all get/query vulnerabilities requests, and allow that state to be updated externally.

Coming from a Java background I also find allowing module-external, direct variable access to be not great, and would rather refactor this whole thing into a class with a method that returns a singleton instance of the class. But 1) I'd need to discuss the startup/instantiation implications of that with the rest of our feature group first, 2) that will break the pattern of the other wrappers (but may be worth it), and 3) that is really outside the scope of this change.

File: tests/unit/anchore_engine/clients/test_grype_wrapper.py
Comment (dspalmer99) [reply tests/unit/anchore_engine/clients/test_grype_wrapper.py:227]: Maybe, but as I said in another comment, I'd rather do something like make the whole module a singleton class in that case. Which I am not opposed to doing (and would probably prefer) but I'd want to discuss it with the group in our sync first, and in any case it is outside the scope of this pr.

File: anchore_engine/clients/grype_wrapper.py
Comment (alfredodeza) [reply anchore_engine/clients/grype_wrapper.py:302]: I'll make this a nit then: whenever a method or function is doing many different things, or directly responsible for many things, that creates a problem: it is harder to maintain, harder to reason about, and harder to write tests. 
Writing tests, or writing code that needs tests - and tests in general - is _not a burden_. The more tested code (regardless of amount of usage) the better.

File: anchore_engine/clients/grype_wrapper.py
Comment (alfredodeza) [reply anchore_engine/clients/grype_wrapper.py:263]: I'll defer this one to other reviewers, but in general, most (if not all) non-error-producing comments can be labeled as "personal preference". You might be right that this in particular is a preference to me, but this has some core reasons why that is so:

Adding code that might make it obvious _at the expense of_ making it redundant causes others to double check the intent. In this particular case: "why is there an `else` here? isn't there something else that the next block  shouldn't do? is there a return?". I had to pause and check those because it wasn't clear to me. The `else:` here is redundant, and it doesn't make it any more clear.

File: anchore_engine/clients/grype_wrapper.py
Comment (alfredodeza) [reply anchore_engine/clients/grype_wrapper.py:242]: I understand the concerns, but I'll have to push back on the reasoning that this is out of scope. This PR is introducing `globals()`, and I'm not finding a reasonable explanation to why that is a requirement. 
I'm not implying that a complete new mechanism needs to exist to handle singletons or configuration values. There are many different ways to declare variables, keep state, and let that state be shared on all consumers _besides_ using `global()`.

This could be:

1. a top-level variable in a module
2. a read-only implementation of a dictionary (or other data structure), to use as a variable
3. a value that gets passed in from function to function
4. an environment variable

I would gladly welcome any code that breaks away from current production code that implements anti-patterns like `globals()` instead of any other option - including singletons, module-level vars, or any other mechanism similar to what I've suggested. 

This  type of addition is going to be a -1 for me.

**Comment by dspalmer99** [comment]
@alfredodeza Why don't we move this to a Zoom call? There seems to be some disagreement about the scope of this PR, and I think it would be faster to resolve this with a f2f discussion.

**Comment by alfredodeza** [comment]
After some offline discussion with @dspalmer99, it was clear that some sort of mutable variable is required because the destination directory keeps changing. Instead of using `global()` or a function that returns a variable that sets it to `global()` I believe that it is better to have a module level variable that has the same effect.

Regardless of that implementation, the underlying cause to force the use of a variable that keeps changing is because the destination directory is always different. I wasn't part of the discussion on how to implement/architect this, but I think that the global state tracking could be avoided by declaring well-known directory paths for the database destination.

All of my other comments were already addressed except for the problematic `globals()` which I still don't agree is a good path forward - but I'll defer to others closer to this functionality that may have a better understanding to continue with the review/approval/merge process.

**Comment by dspalmer99** [comment]
@alfredodeza as per our discussion on Friday, the two most-apparent options to eliminate the use of globals are to either refactor this into a class and make those class variables, or as you said above to keep a consistent db dir name, and rename the new dir with that name, after removing the old (although this adds an additional point of potential failure). For this PR I have moved all references within the wrapper module to getters and setters (only the setters require those variables to use the global modifier). As this PR did not introduce those I do not want to expand the scope on that any further here. Our group working on this discussed this on Friday, and I am going to merge this today.

**Pull request merged**
